[coq] Add support for native-mode compilation.

ejgallego opened the pull request:
We add a `(mode ...)` field to the `(coq.theory ...)` stanza that when
set to `native` will have the Coq compiler to generate and install
"native" objects, that is to say, `.cmxs` libraries containing an
ML-level version of the corresponding Coq module.

As of today, Coq itself does call the `ocamlfind ocaml` to compile the
object files, this restriction may be lifted in the future however it
is not easy as tactics do have access to "Just-In-Time" native
compilation too.

The patch itself should have been straightforward except for a big
problem due to the way Coq outputs native targets and the limitations
of Dune.

By default, `coqc -native-compute on` will generate the OCaml-level
objects files under `.coq-native/`, and the regular `.vo` object in
the build dir. This scheme doesn't work well with the restriction than
a rule can only produce targets in the same directory.

To workaround this limitation, we pass a special flag to Coq so
`.coq-native` is dropped from the output path; at install time we do
set the rules up so indeed the objects will be installed under
`.coq-native` for compatibility. This means this feature will only
work in Coq >= 8.12.1

Moreover, the non-standard location of native objects defeats Coq's
default native include path, so we must pass the additional
non-standard include paths to `coqc`.

Note that native compilation is disabled by default in the `dev`
profile, as to speed up build. Please use the release or a custom
profile if you want to avoid this.

ejgallego commented:
I think the idea for the code is mostly OK [code needs a bit of cleanup but nothing serious]

However there is quite a hairy question in terms of when to enable the native mode.

Compiling with `(mode native)` is usually very expensive, for example we don't want to do that when developing Coq except for a very few cases.

The approach taken in the makefiles is actually to generate that information at configure-time. That is to say, if `-native-compiler no` is passed to `configure` then the native targets won't be generated.

Unfortunately we cannot do that in dune so easily, as that would mean targets for `coqc` depend on an autogenerated file. We could disable this when the profile is `dev`, but actually that would forbid a few useful cases too. I dunno what to do then; need to thing a bit more.

Zimmi48 commented:
Providing different dev profiles seems reasonable to me.

ejgallego commented:
> Providing different dev profiles seems reasonable to me.

Still there is a bad interaction of profiles with Coq's own configure, in the sense that a profile may have native enabled however configure may disable it; then the native objects will fail to be produced and the rule will fail.

This is indeed a consequence of build rules being dependent on `configure` parameters; you could argue that this stuff should not be controlled by configured parameters but by the build system but I'm afraid Coq upstream will loudly complain about it.

Zimmi48 commented:
Well, the plan is to remove configure eventually, isn't it? I agree the situation might be a bit messy in the meantime.

ejgallego commented:
> Well, the plan is to remove configure eventually, isn't it? I agree the situation might be a bit messy in the meantime.

For sure I'd like to reduce configure a lot, but precisely the options about enabling / disabling components of the trusted base may be ones people want still to see in.

I think we can move forward with this using profiles, however there could be some way for users to break the build doing some configure hacks.

Zimmi48 commented:
If we really reduce configure a lot and manage to make it fully optional, then it could be the task of configure to generate the right profile according to users' wishes.

ejgallego commented:
This is certainly experimental and needs documentation, tests, and quite a bit of refactoring, but I'm marking the PR ready for review as IMHO it won't change much. I hope to have it fully ready soon.

The model for native and Coq is still quite in flux, but this should work for the current use cases.

rgrinberg commented:
Loos good to me. Just needs rebase + docs + test and it should be ready to merge. Nice work.

ejgallego commented:
> Loos good to me. Just needs rebase + docs + test and it should be ready to merge. Nice work.

Unfortunately this doesn't fully work in more complex cases due to Coq not liking the local lib outputting native objects in `./` vs installed libs having them in `./.coq_native`. Moreover, `coqdep` is buggy and doesn't reflect the dependency of files on the native modules so adding a rule that would copy the `.cmxs` to `.coq_native` won't work unless we amend coqdep output.

Let me see what can I do.

ejgallego commented:
Ok, new version; this requires a tweak in Coq (c.f. https://github.com/coq/coq/pull/13193 ) but otherwise I'm happy.

The mess of having to deviate from the standard setup is not trivial, but we could do worse.

Code for sure can use a bit of refactoring.

ejgallego commented on src/dune_rules/coq_rules.ml:
```diff
@@ -241,7 +311,7 @@ let parse_coqdep ~dir ~(boot_type : Bootstrap.t) ~coq_module
 
 let deps_of ~dir ~boot_type coq_module =
-  let stdout_to = Coq_module.obj_file ~obj_dir:dir coq_module Dep in
+  let stdout_to = Coq_module.dep_file ~obj_dir:dir coq_module in
```
I've kept the dep file as its own function due to a planned optimization: we hope to call coqdep once for theory as this will provide a significant speed up, unfortunately coqdep needs both to scan the FS and have the sources in place so the finer granularity is not an advantate.

It would be great to have some support akin `-modules` in OCaml, but that's gonna take some non-trivial changes upstream.

ejgallego commented:
@rgrinberg I am happy now with the version of this patch, so feel free to review / comment. Note the caveats.

I'd like to ship in 2.8 , provided Coq's https://github.com/coq/coq/pull/13193 is merged soon.

rgrinberg commented:
The tests are failing. I assume it's because we need the coq PR to be merged first?

The DCO needs to be added to some commits  as well

ejgallego commented:
> The tests are failing. I assume it's because we need the coq PR to be merged first?

Yes, the test will only work for Coq 8.12.1 , so indeed I wonder how to make these tests conditional?

> The DCO needs to be added to some commits as well

Ups, sorry for that.

rgrinberg commented:
I suppose the usual way we make the tests conditional based on the OCaml version.

`(enabled_if (%{coq_version} >= 8.12.1)`

ejgallego commented:
> I suppose the usual way we make the tests conditional based on the OCaml version.
> 
> `(enabled_if (%{coq_version} >= 8.12.1)`

This is something I wanted to try, but can the `%{coq:version}` variable be a target of a rule? [As when Coq itself is in scope, dune has to build Coq first]

rgrinberg commented:
> This is something I wanted to try, but can the %{coq:version} variable be a target of a rule? [As when Coq itself is in scope, dune has to build Coq first]

Good question. It cannot. However, I don't think this completely necessary. When `coq` is part of the project, we could simply detect the version outside the rules, using the same rules as we do for `dune subst`. What do you think?

ejgallego commented:
> Good question. It cannot. However, I don't think this completely necessary. When `coq` is part of the project, we could simply detect the version outside the rules, using the same rules as we do for `dune subst`. What do you think?

Yeah that could work; on the other hand I'm not so sure I'd like to expose `coq_version` to users; at first I thought I'd be a nice idea, so people could do stuff such as `(copy foo_%{coq:version} foo.v)` etc... but as of today, it seems to me like opening a can of worms a bit.

Version-dependent projects [we have a few in Coq's CI] have been so far a big pain to maintain, as opposed to version-specific branches, so I'd say that Dune can just wait for 8.12.1 and test against that version.

rgrinberg commented:
@ejgallego the deadline for making it to 2.8 is this week. Otherwise we'll have to delay this until 2.9

ejgallego commented:
> @ejgallego the deadline for making it to 2.8 is this week. Otherwise we'll have to delay this until 2.9

Should work, Coq 8.12.1 landed in OPAM a couple of days ago. Will update, thanks for the reminder.

ejgallego commented:
I still think this is useful for some users, but clearly experimental.

A very large discussion about Coq and native happened upstream, with main results:
- for 8.13 , build systems will always build native packages if the configure option was set
- for 8.14, a separate `coqnative` binary will be introduced, that does allow to turn `.vo` files into `.cmxs`, this is the long term plan an IMHO a much better fit for what we have here.

Supporting 8.13 properly is a bit of an art as 3 options to the configure script are possible, but let's hear from users.

rgrinberg commented:
> for 8.14, a separate coqnative binary will be introduced, that does allow to turn .vo files into .cmxs, this is the long term plan an IMHO a much better fit for what we have here.

Sounds like a much better approach. I'm fine with merging this as is or waiting until 8.14 until coqnative.

ejgallego commented:
> I'm fine with merging this as is or waiting until 8.14 until coqnative.

I suggest we merge, as there will be a few users of this feature and IMO their feedback would be very valuable when implementing the native setup for `(using coq 1.0)`.

It will also make dune-based OPAM packages for Coq a reality, allowing us to (finally!) remove the makefile system, if we don't find any other issue.

rgrinberg commented on src/dune_rules/coq_rules.ml:
```diff
@@ -19,6 +20,13 @@ module Util = struct
         let info = Lib.info t in
         Lib_info.src_dir info)
 
+  let native_paths ts =
```
I think this function needs a better name. To me, the name it implies it wants the directories of `.cmx` files, but that's not what it's doing.

ejgallego replied:
Indeed, that's used to tell Coq where to look for the -I flags of the coq libraries that are needed, in this case `coq.kernel`. So maybe `coq_native_mllibs_include_paths` ?

rgrinberg replied:
I see. Something like `public_cmi_dirs` sounds enough to me.

ejgallego replied:
Used `coq_nativelib_cmi_dirs` , as we will at some point rename this OCaml lib to `coq.nativelib`, thanks!

rgrinberg commented:
Looks good. I just tweaked the error handling in coq_native_flags not to raise too early.

rgrinberg commented:
Looks like git isn't letting me push to this branch 🤔

ejgallego commented:
> Looks like git isn't letting me push to this branch

Umm, for some reason it was protected ? I dunno, I never had to set that option manually, sorry for that, should work now.

And thanks for the review / fix!

rgrinberg reviewed (approved):
Alright, this looks ready to merge.

ejgallego commented:
> Alright, this looks ready to merge.

You still couldn't push?

rgrinberg commented:
I think I managed. I see my commit in that list

ejgallego commented:
Thanks a lot @rgrinberg , renamed the function, and added a small tweak to make the native `on/off` flag fully explicit as this will be more robust for Coq 8.13 . We could handle Coq's configure setting better, but as in 8.14 `coqnative` will be a different output command it is IMHO not worth it.

ejgallego commented:
@rgrinberg leaving to you to click the merge button due to the 2 new commits.

Zimmi48 commented on doc/dune-files.rst:
```diff
@@ -1668,6 +1669,17 @@ The stanza will build all ``.v`` files on the given directory. The semantics of
   way. Since Coq's lang version ``0.2``.
+- you can enable the production of Coq's native compiler object files
+  by setting ``<coq_native_mode>`` to ``native``, this will pass
+  ``-native-compiler on`` to Coq and install the corresponding object
```
Sorry for only noticing this now, but you probably meant `yes` instead of `on`.

ejgallego replied:
Both are accepted, right? I can switch to `yes/no` if you want.

Zimmi48 replied:
I don't know / am not aware. I said this based on the output of `coqtop -help`.

ejgallego replied:
Yes they are [see for example the discussion on attributes where people advocated for more liberal parsing, but indeed here we see the documentation problem with that apporach]

I hadn't noticed that the documentation says so

Zimmi48 commented on doc/dune-files.rst:
```diff
@@ -1668,6 +1669,17 @@ The stanza will build all ``.v`` files on the given directory. The semantics of
+  is **experimental**, and requires Coq >= 8.12.1; moreover, depending
+  libraries *must* be built with ``(mode native)`` too for this to
+  work; also Coq must be configured to support native
+  compilation. Note that Dune will by explicitly disable output of
```
"will by explicitly" -> "will explicitly"

Zimmi48 reviewed (commented):
Sorry for noticing these small documentation mistakes only now. The documentation says that the supported Coq language versions are `0.1` and `0.2` and forgets to mention `0.3`. The documentation of `(mode ...)` should probably says that it requires version `0.3`.

ejgallego commented:
> Sorry for noticing these small documentation mistakes only now. The documentation says that the supported Coq language versions are `0.1` and `0.2` and forgets to mention `0.3`. The documentation of `(mode ...)` should probably says that it requires version `0.3`.

Thanks a lot for taking the time to review this, I will fix ASAP. Note that the support still remains sketchy, I will add a check soon so if `coqnative` is present and `(coq 0.4)` or above is used the mode flag is deprecated, and instead we have the split set of rules.
